swift/apple/FirezoneKit/Sources/FirezoneKit/Keychain/Keychain.swift

```diff
@@ -6,90 +6,240 @@
 
-enum KeychainError: Error {
+public enum KeychainError: Error {
   case securityError(Status)
+  case appleSecError(call: String, status: Keychain.SecStatus)
+  case nilResultFromAppleSecCall(call: String)
+  case resultFromAppleSecCallIsInvalid(call: String)
+  case unableToFindSavedItem
+  case unableToGetAppGroupIdFromInfoPlist
+  case unableToFormExtensionPath
+  case unableToGetPluginsPath
 }
 
-actor Keychain {
+public actor Keychain {
   private static let account = "Firezone"
+  private let workQueue = DispatchQueue(label: "FirezoneKeychainWorkQueue")
 
-  func store(key: String, data: Data) throws {
-    let query = ([
-      kSecClass: kSecClassGenericPassword,
-      kSecAttrService: getServiceIdentifier(key),
-      kSecAttrAccount: Keychain.account,
-      kSecValueData: data,
-    ] as [CFString: Any]) as CFDictionary
-
-    let status = SecItemAdd(query, nil)
+  public typealias Token = String
+  public typealias PersistentRef = Data
 
-    if status == Status.duplicateItem {
-      try update(key: key, data: data)
-    } else if status != Status.success {
-      throw securityError(status)
-    }
+  public struct TokenAttributes {
+    let authURLString: String
+    let actorName: String
   }
 
-  func update(key: String, data: Data) throws {
-    let query = ([
-      kSecClass: kSecClassGenericPassword,
-      kSecAttrService: getServiceIdentifier(key),
-      kSecAttrAccount: Keychain.account,
-    ] as [CFString: Any]) as CFDictionary
-
-    let updatedData = [kSecValueData: data] as CFDictionary
+  public enum SecStatus: Equatable {
+    case status(Status)
+    case unknownStatus(OSStatus)
 
-    let status = SecItemUpdate(query, updatedData)
+    init(_ osStatus: OSStatus) {
+      if let status = Status(rawValue: osStatus) {
+        self = .status(status)
+      } else {
+        self = .unknownStatus(osStatus)
+      }
+    }
 
-    if status != Status.success {
-      throw securityError(status)
+    var isSuccess: Bool {
+      return self == .status(.success)
     }
   }
 
-  func load(key: String) throws -> Data? {
-    let query = ([
-      kSecClass: kSecClassGenericPassword,
-      kSecAttrService: getServiceIdentifier(key),
-      kSecAttrAccount: Keychain.account,
-      kSecReturnData: kCFBooleanTrue!,
-      kSecMatchLimit: kSecMatchLimitOne,
-    ] as [CFString: Any]) as CFDictionary
-
-    var data: AnyObject?
+  public init() {
+  }
 
-    let status = SecItemCopyMatching(query, &data)
+  func store(token: Token, tokenAttributes: TokenAttributes) async throws -> PersistentRef {
+    #if os(iOS)
+    let query = [
+      // Common for both iOS and macOS:
+      kSecClass: kSecClassGenericPassword,
+      kSecAttrLabel: "Firezone access token (\(tokenAttributes.actorName))",
+      kSecAttrDescription: "Firezone access token",
+      kSecAttrService: tokenAttributes.authURLString,
+      kSecAttrAccount: "\(tokenAttributes.actorName): \(UUID().uuidString)", // The UUID uniquifies this item in the keychain
```

Comment:
typo?

```suggestion
      kSecAttrAccount: "\(tokenAttributes.actorName): \(UUID().uuidString)", // The UUID uniquely identifies this item in the keychain
```

Reply:
I did mean to say "uniquifies", in the sense that "make this item unique, so that it can be reliably saved to the keychain". Is that a correct usage of the term "uniquifies"?

To clarify: The keychain can have only one item with the same group+service+account combination. So in case the same actorName is used by two users, there can be a collision, and possibly some undeterministic behaviour in the keychain, so to make the group+service+account unique, we say account = actorName+UUID. We don't use or keep track of the UUID.